src/mpid/ch4/shm/posix/release_gather/release_gather.c

```diff
@@ -267,6 +268,7 @@ int MPIDI_POSIX_mpi_release_gather_comm_init(MPIR_Comm * comm_ptr,
             } else {
                 /* More shm can be created, update the shared counter */
                 MPL_atomic_fetch_add_uint64(MPIDI_POSIX_shm_limit_counter, memory_to_be_allocated);
+                RELEASE_GATHER_FIELD(comm_ptr, alloc_shm_size) += memory_to_be_allocated;
```

Comment:
The shared memory is to be allocated at comm init time once only, right? So you don't need `+=`. In fact, `alloc_shm_size` is the same as `memory_to_be_allocated`. So how about rename `memory_to_be_allocated` the same as `alloc_shm_size` and here we simply store it to the `comm_ptr`?

Reply:
Actually, you need `+=` because `MPI_Reduce` and `MPI_Bcast` both can allocate shm memory at init stage. So `alloc_shm_size` should be added separately if we call `MPI_Reduce` first and then `MPI_Bcast`. 
I prefer `alloc_shm_size` because `alloc_shm_size` means the total shm mem have been allocated instead of to-be allocated.

> here we simply store it to the comm_ptr?
Do you mean put `alloc_shm_size` to `MPIR_Comm` structure?

Reply:
@KaimingOuyang I am confused. How (or when) is `MPIDI_POSIX_mpi_release_gather_comm_init` being called?

Reply:
`MPIDI_POSIX_mpi_release_gather_comm_init` is called every time you call `MPI_Reduce` or `MPI_Bcast`, but shm memory is only allocated at initial call.

Reply:
I see. So where do we de-allocate the shared memory?  I think we should keep the size of bcast buffer and reduce buffer separately and decrease the counter limit at the same time of shm de-allocation.

Reply:
> where do we de-allocate the shared memory?

In function `MPIDI_POSIX_mpi_release_gather_comm_free`

> I think we should keep the size of bcast buffer and reduce buffer separately and decrease the counter limit at the same time of shm de-allocation

My current code actually does this. bcast and reduce buffer are dealt with separately already, and overall allocated shm mem (including bcast and reduce buffers) associated with a comm is freed at the same time in `MPIDI_POSIX_mpi_release_gather_comm_free`

Reply:
Not exactly. What I suggested is something like this:
```
MPIDU_shm_alloc(comm_ptr, flags_shm_size, (void **) ...);
RELEASE_GATHER_FIELD(comm_ptr, flags_shm_size) = flags_shm_size; 
MPL_atomic_fetch_add(MPIDI_POSIX_shm_limit_count, flags_shm_size);
```
Then at free time:
```
MPIDU_shm_free(...);
MPL_atomic_fetch_sub(MPIDI_POSIX_shm_limit_count, RELEASE_GATHER_FIELD(comm_ptr, flags_shm_size));
```

The idea is each shm_alloc or free is accounted together so we don't need explain the code.

Reply:
would this increase extra atomic operations?

Reply:
The same during init since we are init separately for bcast and reduce anyway. It adds a couple more during comm_free (I think) but we don't care about the overhead during comm_free. And I think the atomic overhead is not of concern even if do it in every individual collective operations since the operation itself will use many atomics anyway so it is not that significant.

Reply:
sounds good. I can change it.